README


This archive contains the distribution Text-vCard,
version 2.12:

  a package to edit and create a single vCard (RFC 2426)

This software is copyright (c) 2012 by Leo Lapworth.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Perl extension Text::vCard.

2.11 Thu 21 May 2012 08:48:46 BST
	- Merge POD cleanup and improvements (szabgab)

2.10 Tue 11 Jan 2011 21:50:02 GMT
    - Merged in quote-printed code (Christian Varga)
    - Cleaned up the tests and code for quote-printed

2.09 Thu 19 Aug 2010 16:08:20 BST
    - Add set_encoding() for export - Stefan

2.08 Tue Jul 13 08:47:58 BST 2010
    - Doc fix (Blair Christensen)

2.07 Wed 28 Jan 2010 22:16:05 GMT
    - Fix bug in get_simple_type() (Philip A. Prindeville)
    - Simplify Makefile.PL and deprecate unused fields (Philip A. Prindeville)

2.06 Sun 24 Jan 2010 10:02:01 GMT
    - Doc typos (Philip A. Prindeville)
    - Add get_simple_type() and import_data() (Philip A. Prindeville)
    - Doc typos (Philip A. Prindeville)

2.05 Fri  8 Jan 2010 22:28:50 GMT
    - Put back changes which were missed out

2.04 Fri  8 Jan 2010 22:00:20 GMT
    - Improvements to Makefile.PL and require 5.6 - thanks cpanservice
    - Switch to git repo
    - Doc patch

2.03 Sun Mar 16 10:10:54 GMT 2008
	- added 'moniker' and marked 'name' as depreciated

2.02 Sat Mar 15 17:15:16 GMT 2008
	- http://rt.cpan.org/Ticket/Display.html?id=34044 
	  Clean up tests so they pass in perl 5.10
      thanks to Neil Williams

2.01 Sun Sep 2 2007
	- Stop empty TYPE= being set on export when not needed
	thanks to Vadim for the suggestion

2.00 Sat Oct 21 2006
	- export now gives 'TYPE=x,y' instead of the incorrect 'x;y'
	
1.99 Wed 30th Aug 2006
	- Added 'quoted-printable' fix [rt.cpan.org #17790]
	this requires MIME::QuotedPrint

1.98 Mon Jul 17 2006
	- Test fix

1.97 Sun Jul  2 2006
	- removed test because I don't think Test::More::is_deaply
	works correctly on: perl 5.8.0 on Solaris x86 9

1.96 Sun 3rd July 2005
	- Removed String::ShellQuote - forgot to PREREQ and not using it

1.95 Sun 26th June 2005
	- Made export() work for address book, still missing encoding
	  though.
	- Added basic test for export()
	- A few little clean up things

1.94 Sun 24th April 2005
        - Made check for 'VCARD' tag case insensative,
        probably won't help until Text::vFile::asData does as well.

1.93 Mon 31st Jan 2005
	- Restructured object so all nodes are stored in a hashref
	  $self->{nodes} so I can get get_group to work, all other
	  methods updated to reflect this.
	- added get_group() to vcard

1.92	- Added group() to Node and extract that info
	  from the vcard if it is there.

1.91
	- Doc bug, said 'tels', should have been 'tel'.

1.9 
	- removed import() from docs as not used now
	- changed Node so excepts data with less than field order elements 
	e.g. N, accepts 'Jon;Smith', As well as 'Jon;Smith;Dr;' this means
	we work better according to the RFC
	- ORGs' 'unit' method requires and returns an array ref as it is
	unlimited in the number of elements it has

1.8 Fri 15th Oct 2004
	- Applied patch from Aaron Cope for version stuff, can
	  now access it as lowercase version().

1.7 Wed 13th Oct 2004
	- remembered to update module versions as well as makefile!

1.6 Wed 13th Oct 2004
	- removed import() and made anonsub again
	- removed version() as was conflicting with $VERSION, 
	  info can be accessed through other methods in anycase

1.5 Wed  Oct 6th 2004
	- Change to support asData change for 'type=home;type=pref'
	- Change to support escaped ; (\;) in data - 'note' now works

1.4 Tue  Oct 5th 2004
	- Patches from Arron for XML::Generator::vCard
	- put auto methods in sub rather than closure for vCard.pm
	- change read_file() to scalar read_file() in address

1.3 Tue Oct 5th 2004
	- Added missing File::Slurp dependency
	- Renamed 'element_type' to 'node_type' as well
	- doc updates with help from binary

1.2 Sat Oct 2nd 2004
	- Changed so can parse group.NODE, the group is ignored for now
	- Added test for apple vcard version 3
	- emailed asData author about small buy with types (home/pref)
	- added autogen methods; $vcard->fn(), $vcard->fullname($value)

1.1 Wed Sept 15th 2003
	- Lots of tests added
	- Restructured (it's Tom's fault) so we use Addressbook as the main
	interface, and Text::vCard is a single card, in the Addressbook object
	- Renamed add_type to add_types
	- Renamed remove_type to remove_types
	- Created Text::vCard::Node to replace all Text::vCard::Part*
	- Dropped requirement for Text::vFile
	- Created new and load methods, dropped iterator
	- Renamed update_value to export_data and changed returned value

1.0  Wed July  23rd 11:18:00 2003
	- More doc fixing
	- Fixed bug in UID and TZ
	- test 02-evolution.t ready for Jay's 'type' fix to support v2.1
	- test 03-phone_v1.t ready for Jay's 'type' fix to support v1.0

0.9  Sat July  07 10:19:00 2003
	- Fixed major doc error!


0.8  Sat May  17 21:19:00 2003
	- Added ability to specify types and tests
	- Altered add_type() and remove_type() to accept array ref as well

0.7  Tue May  13 08:00:00 2003
	- Added emails, tels, lables. Started Binary API

0.6  Fri May  9 18:10:00 2003
	- More accessor methods added, added the version to everything as well

0.5  Fri May  9 08:30:00 2003
	- More accessor methods added, docs and tests

0.4  Wed May  7 14:48:20 2003
	- Lots of new stuff, it's starting to take shape

0.3  Wed May  7 10:20:20 2003
	- Cleaned up the docs, structure (now in lib/) and Makefile.PL

0.2  Tue May  6 12:30:20 2003
	- Release for Jay to see

0.1  Mon May  5 11:50:20 2003
	- original version



lib/Text/vCard.pm
package Text::vCard;

use 5.006;
use Carp;
use strict;
use warnings;
use File::Slurp;
use Text::vFile::asData;
use Text::vCard::Node;

# See this module for your basic parser functions
use base qw(Text::vFile::asData);
use vars qw ($VERSION %lookup %node_aliases @simple);
$VERSION = '2.11';

# If the node's data does not break down use this
my @default_field = qw(value);

# If it does use these
%lookup = (
    'ADR' => [
        'po_box', 'extended',  'street', 'city',
        'region', 'post_code', 'country'
    ],
    'N'   => [ 'family', 'given', 'middle', 'prefixes', 'suffixes' ],
    'GEO' => [ 'lat',    'long' ],
    'ORG' => [ 'name',   'unit' ],
);

%node_aliases = (
    'FULLNAME'  => 'FN',
    'BIRTHDAY'  => 'BDAY',
    'TIMEZONE'  => 'TZ',
    'PHONES'    => 'TEL',
    'ADDRESSES' => 'ADR',
    'NAME'      => 'N',      # To be deprecated as clashes with RFC
    'MONIKER'   => 'N',
);

# Generate all our simple methods
@simple
    = qw(FN BDAY MAILER TZ TITLE ROLE NOTE PRODID REV SORT-STRING UID URL CLASS FULLNAME BIRTHDAY TZ NAME EMAIL NICKNAME PHOTO);

# Now we want lowercase as well
map { push( @simple, lc($_) ) } @simple;

# Generate the methods
{
    no strict 'refs';
    no warnings 'redefine';

    # 'version' handled separately
    # to prevent conflict with ExtUtils::MakeMaker
    # and $VERSION
    for my $node ( @simple, "version" ) {
        *$node = sub {
            my ( $self, $value ) = @_;

            # See if we have it already
            my $nodes = $self->get($node);
            if ( !defined $nodes && $value ) {

                # Add it as a node if not exists and there is a value
                $self->add_node( { 'node_type' => $node, } );

                # Get it out again
                $nodes = $self->get($node);
            }

            if ( scalar($nodes) && $value ) {

                # Set it
                $nodes->[0]->value($value);
            }

            return $nodes->[0]->value() if scalar($nodes);
            return undef;
            }
    }
}

=head1 NAME

Text::vCard - a package to edit and create a single vCard (RFC 2426) 

=head1 WARNING

To handle a whole addressbook with several vCard entries in it, you probably
want to start with L<Text::vCard::Addressbook>, then this module.

This is not backwards compatable with 1.0 or earlier versions! 

Version 1.1 was a complete rewrite/restructure, this should not happen again.

=head1 SYNOPSIS

  use Text::vCard;
  my $cards = Text::vCard->new({
	'asData_node' => $objects_node_from_asData,
  });

=head1 DESCRIPTION

A vCard is an electronic business card. 

This package is for a single vCard (person / record / set of address information).
It provides an API to editing and creating vCards, or supplied a specific piece
of the Text::vFile::asData results it generates a vCard with that content.

You should really use L<Text::vCard::Addressbook> as this handles creating
vCards from an existing file for you.

=head1 METHODS

=head2 new()

  use Text::vCard;

  my $new_vcard = Text::vCard->new();
  
  my $existing_vcard = Text::vCard->new({
	'asData_node' => $objects_node_from_asData,
  });
  
=cut

sub new {
    my ( $proto, $conf ) = @_;
    my $class = ref($proto) || $proto;
    my $self = {};

    bless( $self, $class );

    my %nodes;
    $self->{nodes} = \%nodes;

    if ( defined $conf->{'asData_node'} ) {

        # Have a vcard data node being passed in
        while ( my ( $node_type, $data ) = each %{ $conf->{'asData_node'} } )
        {
            my $group;
            if ( $node_type =~ /\./ ) {

                # Version 3.0 supports group types, we do not
                # so remove everything before '.'
                ( $group, $node_type ) = $node_type =~ /(.+)\.(.*)/;
            }

            # Deal with each type (ADR, FN, TEL etc)
            $self->_add_node(
                {   'node_type' => $node_type,
                    'data'      => $data,
                    'group'     => $group,
                }
            );
        }
    }    # else we're creating a new vCard

    return $self;
}

=head2 add_node()

my $address = $vcard->add_node({
	'node_type' => 'ADR',
});

This creates a new address (a L<Text::vCard::Node> object) in the vCard
which you can then call the address methods on. See below for what options are available.

The node_type parameter must conform to the vCard spec format (e.g. ADR not address)

=cut

sub add_node {
    my ( $self, $conf ) = @_;
    croak 'Must supply a node_type'
        unless defined $conf && defined $conf->{'node_type'};
    unless ( defined $conf->{data} ) {
        my %empty;
        my @data = ( \%empty );
        $conf->{'data'} = \@data;
    }

    $self->_add_node($conf);
}

=head2 get()

The following method allows you to extract the contents from the vCard.

  # get all elements
  $nodes = $vcard->get('tel');

  # Just get the home address
  my $nodes = $vcard->get({
	'node_type' => 'addresses',
	'types' => 'home',
  });
  
  # get all phone number that matches serveral types
  my @types = qw(work home);
  my $nodes = $vcard->get({
	'node_type' => 'tel',
	'types' => \@types,
  });
 
Either an array or array ref is returned, containing L<Text::vCard::Node> objects.
If there are no results of 'node_type' undef is returned.

Supplied with a scalar or an array ref the methods
return a list of nodes of a type, where relevant. If any
of the elements is the prefered element it will be
returned as the first element of the list.

=cut

sub get {
    my ( $self, $conf ) = @_;
    carp "You did not supply an element type" unless defined $conf;
    if ( ref($conf) eq 'HASH' ) {
        return $self->get_of_type( $conf->{'node_type'}, $conf->{'types'} )
            if defined $conf->{'types'};
        return $self->get_of_type( $conf->{'node_type'} );
    } else {
        return $self->get_of_type($conf);
    }
}

=head2 get_simple_type()

The following method is a convenience wrapper for accessing simple elements.

  $value = $vcard->get_simple_type('email', ['internet', 'work']);

If multiple elements match, then only the first is returned.  If the object
isn't found, or doesn't have a simple value, then undef is returned.
 
The argument type may be ommitted, it can be a scalar, or it can be an
array reference if multiple types are selected.

=cut

sub get_simple_type {
    my ( $self, $node_type, $types ) = @_;
    carp "You did not supply an element type" unless defined $node_type;

    my %hash = ('node_type', $node_type);
    $hash{'types'} = $types if defined $types;
    my $node = $self->get(\%hash);
    return undef unless $node && @{$node} > 0 && exists $node->[0]->{'value'};

    $node->[0]->{'value'};
}

=head2 nodes

  my $addresses = $vcard->get({ 'node_type' => 'address' });

  my $first_address = $addresses->[0];
  
  # get the value
  print $first_address->street();

  # set the value
  $first_address->street('Barney Rubble');

  # See if it is part of a group
  if($first_address->group()) {
	print 'Group: ' . $first_address->group();
  }
  
According to the RFC the following 'simple' nodes should only have one element, this is
not enforced by this module, so for example you can have multiple URL's if you wish.

=head2 simple nodes

For simple nodes, you can also access the first node in the following way:

  my $fn = $vcard->fullname();
  # or setting
  $vcard->fullname('new name');

The node will be automatically created if it does not exist and you supplied a value.
undef is returned if the node does not exist. Simple nodes can be called as all upper
or all lowercase method names.

  vCard Spec: 'simple'    Alias
  --------------------    --------
  FN                      fullname
  BDAY                    birthday
  MAILER
  TZ                      timezone
  TITLE 
  ROLE 
  NOTE 
  PRODID 
  REV 
  SORT-STRING 
  UID
  URL 
  CLASS
  EMAIL
  NICKNAME
  PHOTO
  version (lowercase only)
  
=head2 more complex vCard nodes

  vCard Spec    Alias           Methods on object
  ----------    ----------      -----------------
  N             name (depreciated as conflicts with rfc, use moniker)
  N             moniker            'family','given','middle','prefixes','suffixes'
  ADR           addresses       'po_box','extended','street','city','region','post_code','country'
  GEO                           'lat','long'
  TEL           phones
  LABELS
  ORG                           'name','unit' (unit is a special case and will return an array reference)

  my $addresses = $vcard->get({ 'node_type' => 'addresses' });
  foreach my $address (@{$addresses}) {
	print $address->street();
  }

  # Setting values on an address element
  $addresses->[0]->street('The burrows');
  $addresses->[0]->region('Wimbeldon common');

  # Checking an address is a specific type
  $addresses->[0]->is_type('fax');
  $addresses->[0]->add_types('home');
  $addresses->[0]->remove_types('work');

=head2 get_group()

  my $group_name = 'item1';
  my $node_type = 'X-ABLABEL';
  my $of_group = $vcard->get_group($group_name,$node_type);
  foreach my $label (@{$of_group}) {
	print $label->value();
  }

This method takes one or two arguments. The group name
(accessable on any node object by using $node->group() - not
all nodes will have a group, indeed most vcards do not seem
to use it) and optionally the types of node you with to 
have returned.

Either an array or array reference is returned depending
on the calling context, if there are no matches it will
be empty.

=cut

sub get_group {
    my ( $self, $group_name, $node_type ) = @_;
    my @to_return;

    carp "No group name supplied"
        unless defined $group_name
            and $group_name ne '';

    $group_name = lc($group_name);

    if ( defined $node_type && $node_type ne '' ) {

        # After a specific node type
        my $nodes = $self->get($node_type);
        foreach my $node ( @{$nodes} ) {
            push( @to_return, $node ) if $node->group() eq $group_name;
        }
    } else {

        # We want everything from that group
        foreach my $node_loop ( keys %{ $self->{nodes} } ) {

            # Loop through each type
            my $nodes = $self->get($node_loop);
            foreach my $node ( @{$nodes} ) {
                if ( $node->group() ) {
                    push( @to_return, $node )
                        if $node->group() eq $group_name;
                }
            }
        }
    }
    return wantarray ? @to_return : \@to_return;
}

=head1 BINARY METHODS

These methods allow access to what are potentially
binary values such as a photo or sound file.

API still to be finalised.

=head2 photo()

=head2 sound()

=head2 key()

=head2 logo()

=cut

sub DESTROY {
}

=head2 get_lookup

This method is used internally to lookup those nodes which have multiple elements,
e.g. GEO has lat and long, N (name) has family, given, middle etc.

If you wish to extend this package (for custom attributes), overload this method
in your code

  sub my_lookup {
		return \%my_lookup;
  }
  *Text::vCard::get_lookup = \&my_lookup;

This has not been tested yet.

=cut 

sub get_lookup {
    my $self = shift;
    return \%lookup;
}

=head2 get_of_type()

  my $list = $vcard->get_of_type($node_type,\@types);

It is probably easier just to use the get() method, which inturn calls
this method.

=cut

# Used to get the right elements
sub get_of_type {
    my ( $self, $node_type, $types ) = @_;

    # Upper case the name
    $node_type = uc($node_type);

    # See if there is an alias for it
    $node_type = uc( $node_aliases{$node_type} )
        if defined $node_aliases{$node_type};

    return undef unless defined $self->{nodes}->{$node_type};

    if ($types) {

        # After specific types
        my @of_type;
        if ( ref($types) eq 'ARRAY' ) {
            @of_type = @{$types};

            #	print "T A: " . join('-',@{$types}) . "\n";
        } else {
            push( @of_type, $types );

            #	print "T: $types\n";
        }
        my @to_return;
        foreach my $element ( @{ $self->{nodes}->{$node_type} } ) {
            my $check = 1;    # assum ok for now
            foreach my $type (@of_type) {

                # set it as bad if we don't match
                $check = 0 unless $element->is_type($type);
            }
            if ( $check == 1 ) {

                #	print "Adding: $element->street() \n";
                push( @to_return, $element );
            }
        }

        return undef unless scalar(@to_return);

        # Make prefered value first
        @to_return = sort { _sort_prefs($b) <=> _sort_prefs($a) } @to_return;

        return wantarray ? @to_return : \@to_return;

    } else {

        # Return them all
        return wantarray
            ? @{ $self->{nodes}->{$node_type} }
            : $self->{nodes}->{$node_type};
    }
}

sub _sort_prefs {
    my $check = shift;
    if ( $check->is_type('pref') ) {
        return 1;
    } else {
        return 0;
    }
}

# Private method for adding nodes
sub _add_node {
    my ( $self, $conf ) = @_;

    my $value_fields = $self->get_lookup();

    my $node_type = uc( $conf->{node_type} );
    $node_type = $node_aliases{$node_type}
        if defined $node_aliases{$node_type};

    my $field_list;

    if ( defined $value_fields->{$node_type} ) {

        # We know what the field list is
        $field_list = $value_fields->{$node_type};
    } else {

        # No defined fields - use just the 'value' one
        $field_list = \@default_field;
    }
    unless ( defined $self->{nodes}->{$node_type} ) {

        # create space to hold list of node objects
        my @node_list_space;
        $self->{nodes}->{$node_type} = \@node_list_space;
    }
    my $last_node;
    foreach my $node_data ( @{ $conf->{data} } ) {
        my $node_obj = Text::vCard::Node->new(
            {   node_type => $node_type,
                fields    => $field_list,
                data      => $node_data,
                group     => $conf->{group} || '',
            }
        );

        push( @{ $self->{nodes}->{$node_type} }, $node_obj );

        # store the last node so we can return it.
        $last_node = $node_obj;
    }
    return $last_node;
}

=head1 AUTHOR

Leo Lapworth, LLAP@cuckoo.org

=head1 BUGS

None that I'm aware of - export may not encode correctly.

=head1 Repository (git)

http://github.com/ranguard/text-vcard, git://github.com/ranguard/text-vcard.git

=head1 COPYRIGHT

Copyright (c) 2005-2010 Leo Lapworth. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<Text::vCard::Addressbook>, L<Text::vCard::Node>

=cut

1;
